site-src/geps/gep-1058.md

```diff
@@ -0,0 +1,962 @@
+instead of having to embed those settings directly in each Route or RouteRule.
+
+The configurations for Scenario 1 (shown lower on this page), demonstrate two
+ways this technique can be used. In the first case, traffic needs to be routed
+to "Core Services" (`core-svces-ns`) from two different starting points. Using
+route inclusion, the routing configuration for the Core Services are put in a
+child route and then that child route is included in two sections of a parent
+route where they are needed.
+
+Another way route inclusion is used in Scenario 1 is by configuring a
+HTTPRouteFilter, that inserts a header, in the parent route. This results in the
+filter being added to each RouteRule in the child route `store-route`. The child
+route is both shorter and easier to read than it would have been without this
+technique.
+
+This use case and the Delegation use case are solved using the same mechanism
+(route inclusion). They differ only in the intention (i.e. purpose) for using
+that mechanism.
+
+One advantage of allowing Routes to include groups of settings by reference is
+that, when one of those settings needs to be changed, it only has to be changed
+in one place instead of editing all of the relevant Routes. Another advantage is
+that it prevents those settings from diverging across the Routes over time.
+
+The remainder of this document covers the proposed enhancements to the Gateway API.
+
+## Patent/Child Route Relationships
+
+The proposal is to allow the settings in a child route to be combined with a
+section of the parent route’s RouteRule.   A HTTPRouteRule may have either
+BackendRefs or AllowedRoutes set. The effect of this is that a Route can
+delegate to other Routes. These routes will be referred to as the parent and
+child routes. This model would be more consistent with the current Route to
+Gateway binding model (as documented in [GEP 724][gep-724]) than alternatives
+such as "route inclusion". It allows minimal coordination from the team
+implementing the child route while remaining within a scope defined by a gateway
+operator, and centralizes configuration on the Route object, avoiding the need
+for a ReferencePolicy.
+
+### Summary of Changes, Requirements and Constraints: Parent/Child Route Relationships
+
+This section details the changes to the standard to support Parent/Child routes
+and the rules that govern those relationships.
+
+(Note: A separate section covers the rules for how parent and child routes are
+logically combined at runtime.)
+
+Before getting to those details, the following two fragments of configuration
+show how a parent/child route relationship is created. These are based on the
+Scenario 1 diagram shown previously.
+
+```yaml
+#### Parent Route #####
+kind: HTTPRoute
+metadata:
+  name: my-example
+  namespace: prod-gw
+spec:
+  parentRefs:
+    name:  prod-web-gw
+  hostname: www.example.com
+  rules:
+  ...
+    - name: store
+...
+      allowedRoutes:
```

Comment:
This seems extremely complex for what we are doing. First, I would assume the majority of use cases are to a single namespace -- maybe even go as far as to say we should only allow a single namespace, as allowing multiple is problematic from an ownership perspective. We also don't need `namespaces` nesting, since `kinds` is not relevant here since we only allow same-kind.

What that leaves us with is the need to refer to a single namespace. This means everything else is boilerplate, really. We could consider simplifying this entire to something like `routeNamespace: store-ns` even.

I do prefer using `backendRef` anyways ,but if we are going with this approach I think it could be dramatically simplified

Reply:
Yeah as I've been thinking about this more it seems simpler and safer to limit this capability to a single namespace and replace this with a single string instead of the struct. I think I'd originally argued for consistency with Gateway -> Route, but that's very much overkill here.

Reply:
For migration, could you do this:

Pre:  `routeNamespace: before`, route defined in `before`
Transition:  `routeNamespace: before`, route defined in `before` and the same route defined in `after`
Switch over:  `routeNamespace: after`, route defined in `before` and the same route defined in `after`
Complete:  `routeNamespace: after`, route defined in `after`

?

It seems mostly the same flow, except with multiple namespaces you have both child routes 'active' at one time and are relying on conflict resolution to ensure only one is picked at time? 

I am not confident in ^, LMK if that makes sense...

Reply:
That's why we have a list of BackendRefs and weights. 

To migrate you add the second namespace with weight 0, and gradually shift. The parent has full control, and you can't accidentally have a random namespace using same labels mess up, or complicated conflict resolution between namespaces.

Reply:
Even a single namespace allowed, there will be cases, multiple delegate rules referenced, how can we process it?

In istio, we allow specifying the delegate by namespace/name,  there can be only one delegate for a root rule.

Reply:
My original intent in proposing [`AllowedRoutes`](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.AllowedRoutes) was to offer config parity between how Gateways allow routes to bind and how HTTPRoute could similarly allow child routes (as @robscott mentioned above).

**With the added same-kind constraint, we can definitely skip at least the first level.** (Even if we wanted to consider extensibility for e.g. TCPRoute -> HTTPRoute delegation in a future iteration, I don't think _additionally_ adding `AllowedRoutes` as a peer at that time would be a breaking change or backwards-incompatible, and this does reduce boilerplate for the simpler case.)

Using the [`RouteNamespaces`](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.RouteNamespaces) struct directly (which we may want to rename to `AllowedNamespaces` for clarity in the standalone use case?) would support both the single-namespace

**For the simplest case, could we add an additional enum variant `"Namespace"` to [`FromNamespaces`](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io/v1alpha2.FromNamespaces)?** Following that, we could add an additional optional field `name` to `RouteNamespaces`, with similar constraints as `selector`:
> `name` string _(Optional)_
Name must be specified when From is set to “Namespace”. In that case, only Routes in Namespaces matching this Name will be selected by this Gateway (or Route). This field is ignored for other values of “From”.

This would have the benefit of keeping our struct types limited _and_ additionally improving the config UX for simple cases of regular Gateway -> Route ownership delegation.

> Even a single namespace allowed, there will be cases, multiple delegate rules referenced, how can we process it?
>
> In istio, we allow specifying the delegate by namespace/name, there can be only one delegate for a root rule.

I think @hzxuzhonghu's point is valid, even for the single-namespace-only case, and we should offer prescriptive guidance/conformance tests for how this should be handled. I think specifying `name` in a delegation slot leans too much toward requiring coordination across the delegation boundary though, and doesn't offer any meaningful additional constraint (team in delegated namespace could rename any route to match, it's just annoying).